### PR TITLE
Choose project popover

### LINKED ISSUE #991: A method to indicate the current project
## Problem

Some more advanced IDE features require us to know which project we're currently working in to function in a predictable manner. Global search, build & run, git features, etc. At the moment, we have no clear way of doing this.

## Proposal

- [ ] Add a dropdown to the headerbar to manage and select projects
- [ ] Only list files and folders from the selected project in the sidebar
- [ ] Remove items that no longer make sense like "close other folders"

This would automatically close:
* #568

image

### PR DESCRIPTION
**Pull request by jeremypw**
Fixes #991 (partially)
Alternative implementation to #1023, similar to that used for FormatBar.

Open questions, given that the number of open projects is likely to be small compared with number of formats:
* What to do with non-active project folders and documents

A potential future development would be to maintain a list of recent but not loaded projects that could also be displayed in the list.

### CODE REVIEW

**Comment by jeremypw** [comment]
ChooseProjectPopover

**Comment by hanaral** [comment]
From the screenshot it's even better than what I was thinking! I think there may be some debate as to whether it should adapt to the number of projects i.e if >8 then make it scrollable otherwise crop the popover to fit the number of entries
I'm in favour of making it adaptable

**Comment by jeremypw** [comment]
@hanaral Yes, unless/until we display recent but not loaded projects as well as loaded projects the list is unlikely to contain more than 3 projections in my opinion. Trying to work with more than that at once would be unwieldy.

**Comment by hanaral** [comment]
I do still think it would be useful to have it, in case Code is used by other projects that need their tools to scale. 
Apart from that, do you think you could fix the button's UX in this PR too? It just needs to be the same size as the other headerbar button widgets, and use a symbolic bit icon rather than the coloured (although a single colour icon could also work, similarly to the settings popover)

**Comment by jeremypw** [comment]
We need to raise an issue in the icons project for a symbolic (git) project icon. 

I'll fix the button height - I copied the code from the formatbar but I must have missed something.

**Comment by jeremypw** [comment]
CodeHeaderBar

Added margins to match the custom title (FormatBar), use `hexpand = true` to avoid width changes when changing project.

Only problem with this is that it pushes the undo button to the right, next to the FormatBar.

**Comment by hanaral** [comment]
I think it might just be necessary to expand to like 8 letters and then ellipsise the rest, considering it could just show the full name on hover.

**Comment by jeremypw** [comment]
@danrabbit in your issue report #991, what is in the right-hand half of the button - I cannot quite read your illustration.  Is it the current branch name with a popover list of available branches?  Have you decided about closing/hiding/collapsing non-active projects? Should documents from non-active projects remain open?

**Comment by jeremypw** [comment]
@hanaral The button now displays max 24 chars, which should be enough for most repos, and ellipsizes if not. May as well show the full path to the repo in the tooltip as when dealing with forks there may be two repos with the same name.

**Review by danrabbit — DISMISSED** [review]
Working my way through here, this is a good direction! 

I appreciate the search entry since I often have a lot of projects open

One regression I noticed is that I can no longer use global search on projects which aren't using Git

In my mockup I had an idea of maybe a linked button for choosing the git branch, but that idea can wait for a follow up since there is already a way to change the branch.

File: src/Widgets/ChooseProjectButton.vala [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            if (project_entry.active) {
+                active_path = project_entry.project_path;
+            }
+        });
+
+        return active_path;
+    }
+
+    public class ProjectEntry : Gtk.ListBoxRow {
+        public bool active;
```
Comment (danrabbit) [inline src/Widgets/ChooseProjectButton.vala:169]: This needs to be explicitly marked as get/set. That's my bad
```suggestion
        public bool active { get; set; }
```

File: src/Widgets/ChooseProjectButton.vala [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            var project_entry = ((ProjectEntry) child);
+            if (project_entry.active) {
+                active_path = project_entry.project_path;
+            }
+        });
+
+        return active_path;
+    }
+
+    public class ProjectEntry : Gtk.ListBoxRow {
```
[the remark this answers is not in the document]
Comment (jeremypw) [reply src/Widgets/ChooseProjectButton.vala]: Trunuff - done.

File: src/Widgets/ChooseProjectButton.vala [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    public void set_document (Scratch.Services.Document doc) {
+        var path = doc.file.get_path ();
+        project_selection_listbox.get_children ().foreach ((child) => {
+            var project_entry = ((ProjectEntry) child);
+            if (path.has_prefix (project_entry.project_path)) {
+                select_project (project_entry);
+            }
+        });
```
[the remark this answers is not in the document]
Comment (jeremypw) [reply src/Widgets/ChooseProjectButton.vala]: Oops

File: src/Widgets/ChooseProjectButton.vala [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+            halign = Gtk.Align.START
+        };
+        grid.add (img);
+        grid.add (label_widget);
+        add (grid);
+
+        project_selection_listbox = new Gtk.ListBox () {
+            selection_mode = Gtk.SelectionMode.SINGLE
+        };
+        var project_selection_filter = new Gtk.SearchEntry () {
```
[the remark this answers is not in the document]
Comment (jeremypw) [reply src/Widgets/ChooseProjectButton.vala:54]: I dislike variables with a generic name like `listbox` but I'll try to be more concise.

File: src/Widgets/ChooseProjectButton.vala [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+public class Code.ChooseProjectButton : Gtk.ToggleButton {
+    private const string NO_PROJECT_SELECTED = N_("No Project Selected");
+    private Scratch.Services.GitManager manager;
+    private Gtk.Image img;
```
[the remark this answers is not in the document]
Comment (jeremypw) [reply src/Widgets/ChooseProjectButton.vala:22]: Yes, that was just copied over from FormatButton - not needed here.

File: src/MainWindow.vala [hunk 6]
```diff
@@ -1006,22 +994,28 @@ namespace Scratch {
...
+        }
+
+        private string? get_target_path_for_git_actions (Variant? path_variant) {
+             string? path = "";
+             if (path_variant != null) {
+                 path = path_variant.get_string ();
+             }
+
+             if (path == "") { // Happens when keyboard accelerator is used
+                 path = toolbar.active_project_path;
```
[the remark this answers is not in the document]
Comment (jeremypw) [reply src/MainWindow.vala:1007]: But the active project should be chosen by the user, not the project manager - and the user sets their choice in the toolbar?

[the remark this answers is not in the document]
Comment (jeremypw) [reply src/MainWindow.vala:1007]: I guess you are right though, the button should just be a view onto an underlying model/controller and the window should deal with that.

**Comment by jeremypw** [comment]
@danrabbit Do we need to address what to do with non-active projects and documents or can we leave that to another PR?

One problem as it is, is that if you change the active project with the button it immediately reverts back to the original project as soon as the current document is refocussed, because the active project is kept in sync with the current document.

**Review by danrabbit — APPROVED** [review]
Yeah, let's leave that case for follow up. Merging this would unblock a lot of other work already, so I'm happy to get it in and we can iterate :)

**Pull request merged**